src/Raven.Server/Documents/DocumentTombstoneCleaner.cs

```diff
@@ -127,19 +116,71 @@ internal bool ExecuteCleanup()
             }
             finally
             {
-                Monitor.Exit(_locker);
+                _locker.Release();
             }
+
             return true;
         }
 
         public void Dispose()
         {
-            lock (_locker) // so we are sure we aren't running concurrently with the timer
+            _locker.Wait();
+
+            try
             {
                 _disposed = true;
                 _timer?.Dispose();
                 _timer = null;
             }
+            finally
+            {
+                _locker.Release();
+            }
+        }
+
+        private class DeleteTombstonesCommand : TransactionOperationsMerger.MergedTransactionCommand
+        {
+            private readonly Dictionary<string, long> _tombstones;
+            private readonly long _minAllDocsEtag;
+            private readonly DocumentDatabase _database;
+
+            public DeleteTombstonesCommand(Dictionary<string, long> tombstones, long minAllDocsEtag, DocumentDatabase database)
+            {
+                _tombstones = tombstones;
+                _minAllDocsEtag = minAllDocsEtag;
+                _database = database;
+            }
+
+            public override int Execute(DocumentsOperationContext context)
+            {
+                var deletionCount = 0;
+
+                foreach (var tombstone in _tombstones)
+                {
+                    var minTombstoneValue = Math.Min(tombstone.Value, _minAllDocsEtag);
+                    if (minTombstoneValue <= 0)
+                        continue;
+
+                    if (_database.DatabaseShutdown.IsCancellationRequested)
+                        break;
+
+                    deletionCount++;
+
+                    try
+                    {
+                        _database.DocumentsStorage.DeleteTombstonesBefore(tombstone.Key, minTombstoneValue, context);
+                    }
+                    catch (Exception e)
+                    {
```

Comment:
This might be runner with other transactions, shouldn't we allow it to bubble up? Then the merger will split the tx.
And the catch will be in the caller.